### PR TITLE
Add tutorial 2

### LINKED ISSUE #112: Add tutorial problem which uses test group mode
Minimum Spanning Tree problem seems like  a suitable problem for this

### PR DESCRIPTION
**Pull request by jonathanirvings**
Fix #112

### CODE REVIEW

File: docs/tutorials/tutorial_2.rst [hunk 1]
```diff
@@ -0,0 +1,537 @@
+Tutorial 2: Minimum Spanning Tree
+=================================
+
+In this tutorial, you will learn how to write a generator for a simple problem, called **Minimum Spanning Tree** with test groups. We will focus on the general idea of using test groups in tcframe.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - simple -> more complicated
- tcframe -> **tcframe**

File: docs/tutorials/tutorial_2.rst [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+Here is the complete problem statement:
+
+----
+
+| **Minimum Spanning Tree**
+| Time Limit: 1 second
+| Memory Limit: 64 MB
+|
+| **Description**
+| Given a weighted connected graph G consisting of N nodes and M edges. The nodes are numbered from 0 to N - 1. You need to select N - 1 out of those M edges such that the graph remains connected if we only consider the selected edges. What is the minimum sum of the weight of the selected edges that satisfy the above constraint?
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Given a weighted, connected graph G consisting of N nodes and M edges. The nodes are numbered from 0 to N - 1. You need to choose N - 1 out of those M edges such that the graph remains connected if we only consider the chosen edges. What is the minimum possible sum of the weight of the chosen edges?

File: docs/tutorials/tutorial_2.rst [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+|
+| **Input Format**
+| The first line contains two integers N and M separated by a space.
+| M lines follow. Each line contains three integers U, V, and W separated by a space. This line indicates that there is an edge connecting node U and V with weight W.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Each line contains three integers U, V, and W separated by a space, representing an edge connecting node U and V with weight W.

File: docs/tutorials/tutorial_2.rst [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+|
+| **Output Format**
+| A single line containing the minimum sum of the weight of the selected edges that satisfy the above constraint.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: A single line containing the minimum possible sum of the weight of the chosen edges.

File: docs/tutorials/tutorial_2.rst [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+|
+| **Sample Output**
+| ``5``
+|
+| **Constraints for all subtasks**
+| 2 ≤ N ≤ 100,000
+| N - 1 ≤ M ≤ 100,000
+| 0 ≤ U, V < N
+| 1 ≤ W ≤ 1,000
+| Graph is guaranteed to be connected and there is no edge connecting a node to itself
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Split this into two constraints

File: docs/tutorials/tutorial_2.rst [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+|
+| **Subtask 1**
+| M ≤ 20
+|
+| **Subtask 2**
+| All W are equal
+|
+| **Subtask 3**
+| No additional constraint on this subtask
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: No additional constraints

File: docs/tutorials/tutorial_2.rst [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+----
+
+Since the steps of preparing the problem package and writing the solution are similar to the previous tutorial, we will not cover those steps in this tutorial. Instead, we will focus on writing the problem spec and test spec. Let's use ``minimum-spanning-tree`` as the slug for this problem.
+
+----
+
+Writing problem spec
+--------------------
+
+The following is a problem spec that is derived directly from the problem statement. We are using the same private method ``eachElementBetween()`` from the previous tutorial. We are adding another private method ``allAreEqual()``. We are also adding several private methods for checking graph properties, namely ``noSelfLoop()`` and ``isConnected()``. In a graph, self loop is an edge connecting a node to itself. These methods are very modular and we can always reuse them for another problem. However, note that the methods assume the nodes of the graph are 0-based, and several minor changes are needed if the nodes are 1-based.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: "In a graph, self loop is an edge connecting a node to itself." -> this sentence is not coherent in the paragraph, at least it should be in a pair of parentheses after `noSelfLoop()`

File: docs/tutorials/tutorial_2.rst [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+
+Let's focus on the syntax differences compared to the first tutorial. Note that we are using the method ``Constraints()`` to include the global constraints that applies to every subtasks, and we are using ``SubtaskX()`` to include the constraints that applies to only those subtasks. Since ``Subtask3()`` does not have any additional constraints that is not included in the global constraints, we can omit the method.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Ugh... this is a bug in our side. `Subtask3()` should be defined even with an empty body. Filed #114 to track this.

File: docs/tutorials/tutorial_2.rst [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+        bool eachElementBetween(const vector<int>& v, int lo, int hi) {
+            for (int x : v) {
+                if (x < lo || x > hi) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        bool allAreEqual(const vector<int> &v) {
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Please put `&` right after the type, everywhere

File: docs/tutorials/tutorial_2.rst [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+            for (int x : v) {
+                if (x != v[0]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        bool noSelfLoop(const vector<int> &U, const vector<int> &V) {
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Lowercase u and v

File: docs/tutorials/tutorial_2.rst [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+            for (int i = 0; i < (int)U.size(); ++i) {
+                if (U[i] == V[i]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        bool isConnected(const int &N, const vector<int> &U, const vector<int> &V) {
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - int n
- lowercase u and v

File: docs/tutorials/tutorial_2.rst [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+            for (int x : v) {
+                if (x != v[0]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        bool noSelfLoop(const vector<int> &U, const vector<int> &V) {
+            for (int i = 0; i < (int)U.size(); ++i) {
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Just use unsigned (and in some other places)

Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: i++ instead ++i, for consistency, everywhere

File: docs/tutorials/tutorial_2.rst [hunk 13]
```diff
@@ -0,0 +1,537 @@
...
+            return numNodesVisited == N;
+        }
+    };
+
+----
+
+Writing test spec
+-----------------
+
+The first difference in the ``TestSpec`` class compared to the first tutorial is in the sample test cases declaration. We need to declare which subtasks will have this sample case as one of their testcases. For the sample input in the problem above, since M ≤ 20 and not all W are equal, then the sample input will be included in subtask 1 and subtask 3.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: testcases -> test cases

File: docs/tutorials/tutorial_2.rst [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+            U.push_back(u);
+            V.push_back(v);
+        }
+    }
+
+Now, let us move on to the creating test cases itself. The actual test cases for a problem involving test groups are challenging. The first step is to draw the venn diagram of the subtasks.
+
+The properties of the subtasks of the above problem are :
+
+1. All testcases inside subtask 1 and subtask 2 are also inside subtask 3, and subtask 3 have more testcases that is neither in subtask 1 nor subtask 2.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: testcases -> test cases (everywhere)

File: docs/tutorials/tutorial_2.rst [hunk 15]
```diff
@@ -0,0 +1,537 @@
...
+        randomTree(N, U, V);
+        while (U.size() < M) {
+            int u = rnd.nextInt(0, N - 2);
+            int v = rnd.nextInt(u + 1, N - 1);
+            U.push_back(u);
+            V.push_back(v);
+        }
+    }
+
+Now, let us move on to the creating test cases itself. The actual test cases for a problem involving test groups are challenging. The first step is to draw the venn diagram of the subtasks.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - creating -> creation
- venn -> Venn

File: docs/tutorials/tutorial_2.rst (hunk 14 above)
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - inside -> in (everywhere)
- subtask 3 have more testcases that is -> subtask 3 contains some other test cases that are

File: docs/tutorials/tutorial_2.rst [hunk 16]
```diff
@@ -0,0 +1,537 @@
...
+
+The properties of the subtasks of the above problem are :
+
+1. All testcases inside subtask 1 and subtask 2 are also inside subtask 3, and subtask 3 have more testcases that is neither in subtask 1 nor subtask 2.
+
+2. Some testcases are inside both subtask 1 and subtask 2, and there are also testcases that is in subtask 1 but not in subtask 2, and there are also testcases that is in subtask 2 but not in subtask 2.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: is -> are

File: docs/tutorials/tutorial_2.rst [hunk 17]
```diff
@@ -0,0 +1,537 @@
...
+
+Therefore, the Venn diagram looks like this
+
+.. image:: tutorial_2-venn_diagram.png
+
+We need to define a test group for each of the region in the Venn diagram. Therefore, we will have four testgroups. Let us number it from 1 to 4 in the same order as the diagram above. Therefore, each testgroups will have the following constraints in addition to the global constraints:
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - testgroups -> test groups (everywhere)
- each testgroups -> the test groups

File: docs/tutorials/tutorial_2.rst [hunk 18]
```diff
@@ -0,0 +1,537 @@
...
+
+1. M ≤ 20 and all W are equal
+2. M ≤ 20 and not all W are equal
+3. M > 20 and all W are equal
+4. M > 20 and not all W are equal
+
+Therefore, the test cases for the first test group looks something like this
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Therefore -> For example, ... can be something like this

File: docs/tutorials/tutorial_2.rst [hunk 19]
```diff
@@ -0,0 +1,537 @@
...
+
+        for (int i = 0; i < 5; ++i) {
+            CASE(N = rnd.nextInt(2, 21),
+                 M = rnd.nextInt(N - 1, 20),
+                 randomGraph(N, M, U, V),
+                 randomWeight(M, W));
+        }
+    }
+
+and so on. Note that in the second testgroup, we manually create a small test case where greedily choosing the first N - 1 edges with smallest weight will create a cycle.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Put this directly as a comment in the source code instead. Also, it's better to put comment for every (group of) test case for more clarity, in general

File: docs/tutorials/tutorials.rst [hunk 20]
```diff
@@ -9,3 +9,5 @@ Here, you can find case studies and best practices on how to write your spec fil
    :maxdepth: 1
 
    tutorial_1
+   tutorial_2
+
```
Comment (fushar) [inline docs/tutorials/tutorials.rst]: extra blank line

File: docs/tutorials/tutorial_2.rst (hunk 1 above)
Comment (jonathanirvings) [inline docs/tutorials/tutorial_2.rst]: Done

**Comment by jonathanirvings** [comment]
All done :)

File: docs/tutorials/tutorial_2.rst [hunk 21]
```diff
@@ -0,0 +1,548 @@
...
+        bool noSelfLoop(const vector<int>& u, const vector<int>& v) {
+            for (int i = 0; i < u.size(); i++) {
+                if (u[i] == v[i]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        bool isConnected(const int& n, const vector<int>& u, const vector<int>& v) {
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: int n (no point on passing reference for built-in simple types)

File: docs/tutorials/tutorial_2.rst [hunk 22]
```diff
@@ -0,0 +1,548 @@
...
+        U.clear();
+        V.clear();
+        W.clear();
+    }
+
+Before creating the actual test cases, let us create the private helper methods that will help us in creating the actual test cases later. Similar to the private helper methods in ``ProblemSpec``, these methods are very modular and we can always reuse them for another problem.
+
+.. sourcecode:: cpp
+
+    void randomWeight(int M, vector<int>& W, int minW = 1, int maxW = 1000) {
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: camelCase parameter names for all methods here

File: docs/tutorials/tutorial_2.rst [hunk 23]
```diff
@@ -0,0 +1,548 @@
...
+
+        for (int i = 0; i < 5; i++) {
+            CASE(N = rnd.nextInt(2, 21),
+                 M = rnd.nextInt(N - 1, 20),
+                 randomGraph(N, M, U, V),
+                 W.assign(M, rnd.nextInt(1, 1000)));
+        }
+    }
+
+It is a good practice to include the smallest case (M = 1) and the largest case (M = 20) satisfying the test group constraints. Since this test group also include subtask 2, we also need to make sure that all W are equal.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: is also included in subtask 2

File: docs/tutorials/tutorial_2.rst [hunk 24]
```diff
@@ -0,0 +1,548 @@
...
+.. sourcecode:: cpp
+
+    void TestGroup2() {
+        Subtasks({1, 3});
+        
+        CASE(N = 4, M = 4,
+             U = {0, 1, 2, 0},
+             V = {1, 2, 0, 3},
+             W = {1, 1, 1, 2});
+        // We manually create a small test case where greedily choosing
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: Put comment before the case, not after

File: docs/tutorials/tutorial_2.rst [hunk 25]
```diff
@@ -0,0 +1,548 @@
...
+
+----
+
+Writing problem spec
+--------------------
+
+The following is a problem spec that is derived directly from the problem statement. We are using the same private method ``eachElementBetween()`` from the previous tutorial. We are adding another private method ``allAreEqual()``. We are also adding several private methods for checking graph properties, namely ``noSelfLoop()`` (self loop is an edge connecting a node to itself) and ``isConnected()``. These methods are very modular and we can always reuse them for another problem. However, note that the methods assume the nodes of the graph are 0-based, and several minor changes are needed if the nodes are 1-based.
+
+Let's focus on the syntax differences compared to the first tutorial. Note that we are using the method ``Constraints()`` to include the global constraints that applies to every subtasks, and we are using ``SubtaskX()`` to include the constraints that applies to only those subtasks. Since ``Subtask3()`` does not have any additional constraints that is not included in the global constraints, we can simply just leave the method empty.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - that applies -> that apply
- every subtasks -> every subtask
- is not included -> are not included

File: docs/tutorials/tutorial_2.rst [hunk 26]
```diff
@@ -0,0 +1,548 @@
...
+
+        for (int i = 0; i < 5; i++) {
+            CASE(N = rnd.nextInt(2, 21),
+                 M = rnd.nextInt(N - 1, 20),
+                 randomGraph(N, M, U, V),
+                 randomWeight(M, W));
+        }
+    }
+
+Since this test group does not include subtask 2, W must not be equal for all elements. Therefore, the smallest case for this test group is M = 2.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - is not included in subtask 2
- it must not be the case that W are equal for all elements

File: docs/tutorials/tutorial_2.rst [hunk 27]
```diff
@@ -0,0 +1,548 @@
...
+
+The third and fourth test group can be created similarly. You can see the complete code containing the test specifications for the next test groups in the following section.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - The third and fourth test groups
- similarly -> in a similar fashion

File: docs/tutorials/tutorial_2.rst [hunk 28]
```diff
@@ -0,0 +1,548 @@
...
+        }
+    }
+
+Now, let us move on to the creation test cases itself. The actual test cases for a problem involving test groups are challenging. The first step is to draw the Venn diagram of the subtasks.
+
+The properties of the subtasks of the above problem are :
+
+1. All test cases in subtask 1 and subtask 2 are also in subtask 3, and subtask 3 contains some other test cases that are neither in subtask 1 nor subtask 2.
+
+2. Some test cases are in both subtask 1 and subtask 2, and there are also test cases that is in subtask 1 but not in subtask 2, and there are also test cases that are in subtask 2 but not in subtask 2.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: that is -> that are

File: docs/tutorials/tutorial_2.rst [hunk 29]
```diff
@@ -0,0 +1,548 @@
...
+
+Therefore, the Venn diagram looks like this
+
+.. image:: tutorial_2-venn_diagram.png
+
+We need to define a test group for each of the region in the Venn diagram. Therefore, we will have four test groups. Let us number it from 1 to 4 in the same order as the diagram above. Therefore, each test groups will have the following constraints in addition to the global constraints:
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: - each of the regions
- each test groups -> the four test groups

File: docs/tutorials/tutorial_2.rst [hunk 30]
```diff
@@ -0,0 +1,548 @@
...
+            return numNodesVisited == n;
+        }
+    };
+
+----
+
+Writing test spec
+-----------------
+
+The first difference in the ``TestSpec`` class compared to the first tutorial is in the sample test cases declaration. We need to declare which subtasks will have this sample case as one of their test cases. For the sample input in the problem above, since M ≤ 20 and not all W are equal, then the sample input will be included in subtask 1 and subtask 3.
```
Comment (fushar) [inline docs/tutorials/tutorial_2.rst]: sample case, sample input -> sample test case

**Comment by jonathanirvings** [comment]
All done

**Pull request merged**